File: srcs/frontend/local_game/ball.js
export class Ball 
{
    constructor(x, y, radius, speedX, speedY, speedIncrement = 0.5) {
        this.x = x;
        this.y = y;
        this.beginX = x;
        this.beginY = y;
        this.radius = radius;
        this.speedX = speedX;
        this.speedY = speedY;
        this.beginSpeedX = speedX; // Used for resetting the ball
        this.beginSpeedY = speedY; // Used for resetting the ball
        this.speedIncrement = speedIncrement; // Amount to increase speed after each paddle hit
        this.lastPaddletouch = "left";
    }

    draw(ctx) 
    {
        ctx.beginPath();
        ctx.arc(this.x, this.y, this.radius, 0, Math.PI * 2);
        ctx.fillStyle = "#00FF00";
        ctx.fill();
        ctx.closePath();
    }

    getLastPaddleTouch()
    {
        return this.lastPaddletouch;
    }

    update(canvas, leftPaddle, rightPaddle)
    {
        this.x += this.speedX;
        this.y += this.speedY;

        // Collision with the right paddle
        if (this.x + this.radius >= rightPaddle.x && 
            this.y >= rightPaddle.topPaddle &&
            this.y <= rightPaddle.downPaddle) {
            
            // Adjust speedY based on where the ball hits the paddle
            let offset = this.y - (rightPaddle.topPaddle + rightPaddle.height / 2);
            let angle = offset / (rightPaddle.height / 2); // Normalized to range -1 to 1
            
            this.speedX = -this.speedX; // Reverse x direction
            this.speedY = angle * 4; // Change y direction based on hit position
            this.x = rightPaddle.x - this.radius; // Prevent sticking to the paddle

            // Increase speed slightly after each paddle collision
            this.speedX += this.speedX > 0 ? this.speedIncrement : -this.speedIncrement;
            this.speedY += this.speedY > 0 ? this.speedIncrement : -this.speedIncrement;
            this.lastPaddletouch = "right";
        }
        // Collision with the left paddle (similar to right paddle logic)
        else if (this.x - this.radius <= leftPaddle.x + leftPaddle.width &&
                 this.y >= leftPaddle.topPaddle &&
                 this.y <= leftPaddle.downPaddle) {
            
            let offset = this.y - (leftPaddle.topPaddle + leftPaddle.height / 2);
            let angle = offset / (leftPaddle.height / 2);
            
            this.speedX = -this.speedX;
            this.speedY = angle * 4;
            this.x = leftPaddle.x + leftPaddle.width + this.radius;

            this.speedX += this.speedX > 0 ? this.speedIncrement : -this.speedIncrement;
            this.speedY += this.speedY > 0 ? this.speedIncrement : -this.speedIncrement;
            this.lastPaddletouch = "left";
        }
        // Reverse y direction if the ball hits the top or bottom edges of the canvas
        if (this.y + this.radius > canvas.height || this.y - this.radius < 0) {
            this.speedY = -this.speedY;
        }
    }

    // Static sleep function to create a delay
    static sleep(ms)
    {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    // Method to reset the ball's position with a 1-second delay
    async resetPosition(player) {
        // Position de départ légèrement aléatoire autour du centre
        this.x = this.beginX + (Math.random() * 20 - 10); // Décalage entre -10 et 10
        this.y = this.beginY + (Math.random() * 20 - 10);
    
        // Réinitialisation temporaire de la vitesse
        this.speedX = 0;
        this.speedY = 0;
    
        // Pause d'une seconde
        await Ball.sleep(1000);
    
        // Définir la nouvelle direction et vitesse
        const randomSpeedX = Math.random() * 0.5 + 2; // Vitesse aléatoire entre 2 et 2.5
        const randomSpeedY = (Math.random() * 4 - 2); // Vitesse aléatoire entre -2 et 2
    
        // Direction en X déterminée par le joueur qui a marqué
        this.speedX = player === 2 ? randomSpeedX : -randomSpeedX;
    
        // Direction en Y avec un angle aléatoire
        this.speedY = randomSpeedY;
    
    }

}

File: srcs/backend/Project/AppGame/remote_game/ball.py
import random
import asyncio

class Ball:
    def __init__(self, canvas_height, canvas_width, radius=10, speedX=125, speedY=75, speedIncrement=25):
        self.canvas_height = canvas_height
        self.canvas_width = canvas_width
        self.radius = radius
        self.x = canvas_width / 2
        self.y = canvas_height / 2
        self.beginX = self.x
        self.beginY = self.y
        self.speedX = speedX  # Pixels par frame
        self.speedY = speedY  # Pixels par frame
        self.beginSpeedX = speedX
        self.beginSpeedY = speedY
        self.speedIncrement = speedIncrement  # Pixels par frame

    def update(self, leftPaddle, rightPaddle, delta_time):
        self.x += self.speedX * delta_time
        self.y += self.speedY * delta_time

        # Collision avec le paddle droit
        if (self.x + self.radius >= rightPaddle.x and
            self.y >= rightPaddle.y and
            self.y <= rightPaddle.y + rightPaddle.height):

            # Ajuster speedY en fonction de l'endroit où la balle frappe le paddle
            offset = self.y - (rightPaddle.y + rightPaddle.height / 2)
            angle = offset / (rightPaddle.height / 2)  # Normalisé entre -1 et 1

            self.speedX = -self.speedX  # Inverser la direction X
            self.speedY = angle * 4  # Changer la direction Y en fonction de la position d'impact
            self.x = rightPaddle.x - self.radius  # Empêcher la balle de coller au paddle

            # Augmenter légèrement la vitesse après chaque collision avec un paddle
            self.speedX += self.speedIncrement if self.speedX > 0 else -self.speedIncrement
            self.speedY += self.speedIncrement if self.speedY > 0 else -self.speedIncrement

        # Collision avec le paddle gauche
        elif (self.x - self.radius <= leftPaddle.x + leftPaddle.width and
              self.y >= leftPaddle.y and
              self.y <= leftPaddle.y + leftPaddle.height):

            offset = self.y - (leftPaddle.y + leftPaddle.height / 2)
            angle = offset / (leftPaddle.height / 2)

            self.speedX = -self.speedX
            self.speedY = angle * 4
            self.x = leftPaddle.x + leftPaddle.width + self.radius

            self.speedX += self.speedIncrement if self.speedX > 0 else -self.speedIncrement
            self.speedY += self.speedIncrement if self.speedY > 0 else -self.speedIncrement

        # Inverser la direction Y si la balle touche le haut ou le bas du canvas
        if self.y + self.radius > self.canvas_height or self.y - self.radius < 0:
            self.speedY = -self.speedY

    def resetPosition(self, player):
        # Réinitialiser la position initiale et la vitesse de la balle
        self.x = self.beginX
        self.y = self.beginY
        self.speedX = 2
        self.speedY = 2

        # Définir la vitesse et la direction après la pause
        self.speedX = self.beginSpeedX if player == 2 else -self.beginSpeedX
        self.speedY = self.beginSpeedY if random.random() < 0.5 else -self.beginSpeedY


        